### DISCUSSION TITLE
Improve [react] useReducer type inference

### DISCUSSION THREAD

**Discussion by mfp22** [DefinitelyTyped/DefinitelyTyped#63607] [Issues with a @types package]
## The problem

`useReducer` does not infer the `state` parameter from the initial state passed in:

```tsx
const [state, dispatch] = useReducer(
  (state, action: Action) => {
    // ...
  },
  0,
);
```
This should infer `state` as `number`, but it makes it an implicit `any` instead. 

## The cause of the problem

The 3rd parameter of `useReducer` is an `init` argument that returns the type that `state` should be inferred as. Unfortunately, since the 3rd parameter is optional instead of defined with overloads, when you don't supply that parameter TypeScript has no way to know what `state` should be. 

## The solution

There should be an overload that doesn't have the 3rd parameter and infers `state` as the same exact type as the 2nd argument. 

## Repro with solution

https://stackblitz.com/edit/react-ts-gy5gv6?file=App.tsx

[a comment by an automated account is not shown]

**Comment by mfp22** [comment]
Looks like this may have been a TODO already? https://github.com/DefinitelyTyped/DefinitelyTyped/blob/e358cde5edeaa8156bc5b9086dd9f5443c2cadc7/types/react/index.d.ts#L1017
```tsx
// NOTE: without the ReducerState indirection, TypeScript would reduce S to be the most common
    // supertype between the reducer's return type and the initialState (or the initializer's return type),
    // which would prevent autocompletion from ever working.

    // TODO: double-check if this weird overload logic is necessary. It is possible it's either a bug
    // in older versions, or a regression in newer versions of the typescript completion service.
    function useReducer<R extends Reducer<any, any>>(
        reducer: R,
        initialState: ReducerState<R>,
        initializer?: undefined
    ): [ReducerState<R>, Dispatch<ReducerAction<R>>];
```

**Comment by mfp22** [comment]
The concern in that comment seems to be around autocompletion with the returned state object. 

I updated my repro to be an object with property `count`, and when I start to type `count` it autocompletes. 

When I try to return an extra property, TypeScript doesn't warn me. This is normal for TypeScript; the new object _can_ be cast to the same type as `initialState`. If people want type safety for returned objects, they need to specify the return type of the function directly. Like 

```tsx
  const [state, dispatch] = useReducer(
    (state, payload: number): {count: number} => ({
      ...state,
      count: state.count + payload,
      extraProp: 'asdf', // Type '{ count: number; extraProp: string; }' is not assignable to type '{ count: number; }'. Object literal may only specify known properties, and 'extraProp' does not exist in type '{ count: number; }'.(2322)
    }),
    { count: 0 }
  );
```

However, I don't see the mentioned issue of TypeScript creating a superset state. When I mouse over the `state` property returned from my implementation of `useReducer`, it just shows it as the exact same type as `initialState`.

**Comment by mfp22** [comment]
I can make a PR for this if nobody has a problem. I'd just change that last overload to not have the 3rd argument

**Reply by eps1lon** [reply]
It's usually better to do a PR directly to see if the changes is compatible with existing tests. That's usually the hardest part of the discussion: deciding what to break.

**Reply by mfp22** [reply]
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64412

**Comment by mfp22** [comment]
PR: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64412
